src/openrct2-ui/windows/Ride.cpp

```diff
@@ -4176,6 +4179,61 @@ static_assert(std::size(RatingNames) == 6);
                         GameActions::Execute(&rideSetAppearanceAction);
                     }
                     break;
+                }
+                case WIDX_RANDOMISE_VEHICLE_COLOURS:
+                {
+                    auto ride = GetRide(rideId);
+                    if (ride == nullptr)
+                        return;
+
+                    auto rideEntry = ride->GetRideEntry();
+                    if (rideEntry == nullptr)
+                        return;
+
+                    bool allowChangingTrimColour = false;
+                    bool allowChangingTertiaryColour = false;
+                    for (int32_t i = 0; i < ride->num_cars_per_train; i++)
+                    {
+                        uint8_t vehicleTypeIndex = RideEntryGetVehicleAtPosition(ride->subtype, ride->num_cars_per_train, i);
+
+                        if (rideEntry->Cars[vehicleTypeIndex].flags & CAR_ENTRY_FLAG_ENABLE_TRIM_COLOUR)
+                        {
+                            allowChangingTrimColour = true;
+                        }
+                        if (rideEntry->Cars[vehicleTypeIndex].flags & CAR_ENTRY_FLAG_ENABLE_TERTIARY_COLOUR)
+                        {
+                            allowChangingTertiaryColour = true;
+                        }
+                    }
+
+                    int32_t numItems = ride->NumTrains;
+                    if (ride->vehicleColourSettings != VehicleColourSettings::perTrain)
+                        numItems = ride->num_cars_per_train;
+
+                    uint16_t i, colour;
+                    for (i = 0; i < numItems; i++)
+                    {
+                        colour = ScenarioRandMax(COLOUR_NUM_NORMAL);
```

Comment:
Don't use ScenarioRand(Max), this influences the game state which will cause desyncs, use UtilRand() instead.